crates/biome_js_analyze/src/lint/nursery/use_function_components.rs

```diff
@@ -0,0 +1,123 @@
+use biome_analyze::{
+    Ast, Rule, RuleDiagnostic, RuleDomain, RuleSource, context::RuleContext, declare_lint_rule,
+};
+use biome_console::markup;
+use biome_rowan::{AstNode, AstNodeList};
+use biome_rule_options::use_function_components::UseFunctionComponentsOptions;
+
+use crate::react::components::{
+    AnyPotentialReactComponentDeclaration, ReactComponentInfo, ReactComponentKind,
+};
+
+declare_lint_rule! {
+    /// Enforce that components are defined as functions and never as classes.
+    ///
+    /// React in particular allows users to create components using functions or classes.
+    /// However, using functions is generally preferred. This rule enforces the use of function components.
+    ///
+    /// This rule makes an exception for class components that implement `componentDidCatch` because there is
+    /// currently no hook alternative for React. This function is typically used for defining error boundaries.
+    /// It's recommended to define your error boundary once and then reuse it across your application.
+    ///
+    /// If you are using Preact, it has a [`useErrorBoundary`](https://preactjs.com/guide/v10/hooks/#useerrorboundary) hook.
+    ///
+    /// ## Examples
+    ///
+    /// ### Invalid
+    ///
+    /// ```jsx,expect_diagnostic
+    /// class Foo extends React.Component {
+    ///   render() {
+    ///     return (
+    ///       <div>This is a class component.</div>
+    ///     );
+    ///   }
+    /// }
+    /// ```
+    ///
+    /// ### Valid
+    ///
+    /// ```jsx
+    /// function Foo() {
+    ///   return <div>This is a function component.</div>;
+    /// }
+    /// ```
+    ///
+    pub UseFunctionComponents {
+        version: "next",
+        name: "useFunctionComponents",
+        language: "jsx",
+        recommended: false,
+        domains: &[RuleDomain::React],
```

Comment:
Should we call this rule `useReactFunctionComponent`? We can't rename old rules that belong to this domain, but can apply the correct name for new ones. What do you think?

Reply:
Makes sense to me.